File: flow_sdk/builtin/whiteboard.py
"""Whiteboard entity — backed by a folder on disk.

Folder layout::

    ~/agentic-assets/whiteboard/<name>/
        WHITE_BOARD.md     # frontmatter + prose + auto-managed mermaid block
        board.json         # {kind:"excalidraw", version:1, data:<...>}
        thumbnail.svg      # generated on save (exportToSvg)

Creating a Whiteboard entity writes the folder and Markdown main document.
The canvas editor writes board.json and thumbnail.svg. Filesystem rendering,
parsing, and missing-document repair belong to flow_sdk.assets; this Entity
provides the graph-route interface.
"""

from flow_sdk.api.api_types.api_field import APIField, Sharing
from flow_sdk.core import Entity
from flow_sdk.db.drivers.db_base_record import BuiltinEntityType


class Whiteboard(Entity):
    """Whiteboard entity — folder-backed Excalidraw asset."""

    type: str = APIField(default=BuiltinEntityType.WHITEBOARD.value)
    name: str = APIField(default="")
    description: str = APIField(default="")
    asset_ref: str = APIField(default="", sharing=Sharing.PRIVATE)

File: ts_sdk/src/entities/whiteboard.ts
import { APIEntity, registerEntity } from '../APIEntity';
import { dataContext } from '../FlowSync/context';
import { FrontMatterFsRef } from '../fs/FrontMatterFsRef';
import { DockPointerData } from '../models/DockPointer';

/**
 * Whiteboard entity — backed by a WhiteboardRecord on disk
 * (~/.claude/whiteboards/<name>/).
 *
 * Folder layout:
 *   <name>/WHITE_BOARD.md  — frontmatter + prose + auto-managed mermaid block
 *   <name>/board.json      — {kind, version, data} Excalidraw payload
 *   <name>/thumbnail.svg   — regenerated on save
 *
 * Create via:  Whiteboard.createInProject(project, name) — calls the same
 * save() path Skill uses, which POSTs the entity and the server materializes
 * the folder.
 */
@registerEntity
export class Whiteboard extends APIEntity<Whiteboard> {
  static type: string = 'whiteboard';
  static override icon = 'Palette';

  description: string = '';
  asset_ref?: string;
  scope?: string;

  constructor(entity: Partial<Whiteboard> = {}) {
    super(entity);
    this.description = entity.description ?? '';
    this.asset_ref = entity.asset_ref;
    this.scope = entity.scope;
  }

  /** Default open target: the asset editor (URL-first navigate target). */
  override get dockPointer(): DockPointerData {
    return this.assetEditorPointer() ?? this.defaultDockPointer;
  }

  override get editorDockPointer(): DockPointerData {
    return this.assetEditorPointer() ?? super.editorDockPointer;
  }


  /** FrontMatterFsRef for WHITE_BOARD.md. Resolves compute node from dataContext. */
  get doc(): FrontMatterFsRef | null {
    const typeId = dataContext.computeNodeTypeId;
    if (!typeId || !this.asset_ref) return null;
    const mdPath = this.asset_ref.replace(/\/$/, '') + '/WHITE_BOARD.md';
    return new FrontMatterFsRef(mdPath, typeId);
  }

  static async create(name: string, description?: string): Promise<Whiteboard> {
    const wb = new Whiteboard({ name: name.trim(), description: description?.trim() ?? '' });
    return wb.save();
  }

  /** Create in the selected scope, optionally at an exact authorized folder. */
  static async createInProject(
    project: { typeId?: import('../models/TypeId').TypeId } | null,
    name: string,
    destination?: import('../fs/FSRef').FSRefJson,
  ): Promise<Whiteboard> {
    const scopeIds = project?.typeId ? [project.typeId] : [];
    const wb = new Whiteboard({ name: name.trim() });
    return wb.save(scopeIds, destination);
  }
}
